python/ray/serve/batching.py

```diff
@@ -141,20 +150,46 @@ async def _handle_batches(self, func):
             try:
                 # Method call.
                 if self_arg is not None:
-                    results = await func(self_arg, *args, **kwargs)
+                    func_future_or_generator = func(self_arg, *args, **kwargs)
                 # Normal function call.
                 else:
-                    results = await func(*args, **kwargs)
-
-                if len(results) != len(batch):
-                    raise RayServeException(
-                        "Batched function doesn't preserve batch size. "
-                        f"The input list has length {len(batch)} but the "
-                        f"returned list has length {len(results)}."
-                    )
-
-                for i, result in enumerate(results):
-                    futures[i].set_result(result)
+                    func_future_or_generator = func(*args, **kwargs)
+                if isasyncgenfunction(func):
+                    func_generator = func_future_or_generator
+                    async for results in func_generator:
+                        if len(results) != len(batch):
+                            raise RayServeException(
+                                "Batched function doesn't preserve batch size. "
+                                f"The input list has length {len(batch)} but the "
+                                f"returned list has length {len(results)}."
+                            )
+
+                        new_futures = [
+                            get_or_create_event_loop().create_future() for _ in futures
+                        ]
+                        for i, result in enumerate(results):
+                            futures[i].set_result(
+                                _GeneratorIterResponse(result, new_futures[i])
+                            )
+                        futures = new_futures
+
+                    for future in futures:
+                        future.set_result(
+                            _GeneratorIterResponse(None, END_ASYNC_ITER_TOKEN)
```

Comment:
Why do we need this special token? Can we have the future raise `StopIteration` instead?

Reply:
Good idea, I changed this, so the future raises a `StopAsyncIteration` instead.